### PR TITLE
Add support for field name remapping hook

### PR DESCRIPTION
**Pull request by 15cm**
Solves #89. Add support to match field name with map key in `snake_case` or `kebab-case` convention.

### CODE REVIEW

**Comment by camdencheek** [comment]
This could be made more generally useful by allowing users to provide a matching function. Something like `type FieldMatchFunc func(from, to string) bool`

**Comment by 15cm** [comment]
In my opinion, it's too heavy for a user to provide a function in a config and several enumerations should cover all naming styles.

File: mapstructure.go [hunk 1]
```diff
@@ -1064,7 +1070,16 @@ func (d *Decoder) decodeStructFromMap(name string, dataVal, val reflect.Value) e
 					continue
 				}
 
-				if strings.EqualFold(mK, fieldName) {
+				matched := strings.EqualFold(mK, fieldName)
+
+				switch d.config.FieldNameTransFormMethod {
+				case "snake":
+					matched = matched || strings.EqualFold(strings.ReplaceAll(mK, "_", ""), fieldName)
+				case "kabab":
```
Comment (camdencheek) [inline mapstructure.go:1078]: `kabab` => `kebab`

**Comment by mitchellh** [comment]
#250

**Pull request closed without merging**